File: julia/Euler/src/bin/p0014.jl

# project euler: problem 14

module Prob0014

memo = Dict{Int, Int}(1 => 1)

function get_collatz_length(n)
    get!(memo, n) do
        if iseven(n)
            get_collatz_length(n ÷ 2) + 1
        else
            get_collatz_length(3n + 1) + 1
        end
    end
end

function solve_0014(limit::Int = 1_000_000)
    argmax(get_collatz_length, (limit ÷ 2):(limit - 1))
end

end #module

using .Prob0014: solve_0014
export solve_0014

File: rust/src/bin/p0014.rs
// Project Euler: Problem 14

euler::run_solver!(14);

fn solve() -> String {
    compute(1_000_000).to_string()
}

fn compute(thr: i64) -> i64 {
    ((thr / 2)..thr)
        .max_by_key(|&n| get_collatz_length(n))
        .unwrap()
}

fn get_collatz_length(mut n: i64) -> i64 {
    let mut cnt: i64 = 1;
    while n > 1 {
        cnt += 1;
        if n & 1 == 0 {
            n /= 2;
        } else {
            n = 3 * n + 1;
        }
    }
    cnt
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0014_under_1000000() {
        assert_eq!(compute(1_000_000), 837799);
    }
}
